torch_xla/csrc/runtime/ifrt_computation_client.cc

```diff
@@ -0,0 +1,678 @@
+                    ifrt_device, xla::ifrt::MemoryKind()),
+                xla::ifrt::Client::HostBufferSemantics::
+                    kImmutableUntilTransferCompletes,
+                [tensor, timed]() { /* frees tensor and timer */ })
+            .value();
+
+    ComputationClient::DataPtr data =
+        std::make_shared<IfrtData>(tensor->device(), tensor->shape(), buffer);
+    datas.push_back(data);
+  }
+  OutboundDataMetric()->AddSample(total_size);
+  CreateDataHandlesCounter()->AddValue(datas.size());
+
+  return datas;
+}
+
+ComputationClient::DataPtr IfrtComputationClient::TransferShardsToServer(
+    absl::Span<const std::shared_ptr<const TensorSource>> tensor_shards,
+    std::string device, xla::Shape shape, xla::OpSharding sharding) {
+  tsl::profiler::TraceMe activity(
+      "IfrtComputationClient::TransferShardsToServer",
+      tsl::profiler::TraceMeLevel::kInfo);
+  // TODO(jonbolin): Consider using CopyToDevice when sharding is REPLICATED.
+  // We are opting out of CopyToDevice for now due to the synchronization
+  // issues observed in ShardingUtil::InputHandler, but because CopyToDevice
+  // directly copies buffers between devices using ICI, it can be much faster
+  // than transferring from the host to each device.
+  auto data_shards = TransferToServer(tensor_shards);
+  std::vector<tsl::RCReference<xla::ifrt::Array>> arrays;
+  std::vector<xla::ifrt::Shape> shard_shapes;
+  for (auto& shard : data_shards) {
+    auto ifrt_shard = std::dynamic_pointer_cast<IfrtData>(shard);
+    arrays.push_back(ifrt_shard->buffer);
+    shard_shapes.push_back(ifrt_shard->buffer->shape());
+  }
+  xla::ifrt::Shape ifrt_shape(shape.dimensions());
+  xla::ifrt::DeviceList devices_list({client_->addressable_devices().begin(),
+                                      client_->addressable_devices().end()});
+  std::unique_ptr<xla::ifrt::Sharding> ifrt_sharding =
+      xla::ifrt::ConcreteSharding::Create(devices_list, xla::ifrt::MemoryKind(),
+                                          ifrt_shape, shard_shapes);
+  // TODO: Attach HloSharding instead when it is supported
+  // std::unique_ptr<xla::ifrt::Sharding> ifrt_sharding =
+  // xla::ifrt::HloSharding::Create(
+  //   devices_list,
+  //   xla::ifrt::MemoryKind(),
+  //   xla::HloSharding::FromProto(sharding).value()
+  // );
+  tsl::RCReference<xla::ifrt::Array> sharded_array =
+      client_
+          ->AssembleArrayFromSingleDeviceArrays(
+              ifrt_shape, std::move(ifrt_sharding), absl::MakeSpan(arrays),
+              xla::ifrt::ArrayCopySemantics::kAlwaysCopy)
+          .value();
+  return std::make_shared<IfrtData>(device, shape, sharded_array, sharding);
+}
+
+ComputationClient::DataPtr IfrtComputationClient::CopyToDevice(
+    ComputationClient::DataPtr data, std::string dst) {
+  XLA_ERROR() << __FUNCTION__ << " not implemented";
+}
+
+tsl::RCReference<xla::ifrt::Array> IfrtComputationClient::ReplicateShardedData(
```

Comment:
Do we still need this API in Ifrt or could we `CopyToHostBuffer` directly on an Array backed by sharded data?

Reply:
> could we CopyToHostBuffer directly on an Array backed by sharded data?

IIUC this will not work with Arrays sharded over non-addressable devices

Reply:
Oh that's interesting - thanks for the clarification, I had assumed Arrays represented the global array.

Reply:
Arrays can represent _either_ global arrays  or shards of a global array. You can even have both a global and local array pointing to the same underlying data.